--- /mnt/data/sunxi/A20/lichee/linux-3.3/drivers/input/touchscreen/gt9xx.c	2014-09-08 16:55:33.000000000 +0000
+++ drivers/input/touchscreen/gt9xx.c	2016-07-03 19:02:55.000000000 +0000
@@ -366,6 +366,9 @@ static void gtp_touch_down(struct goodix
         
         dprintk(DEBUG_X_Y_INFO,"report data:ID:%d, X:%d, Y:%d, W:%d\n", id, x, y, w);
 
+	y += 5;
+	if( y > 800 )
+	    y = 800;
 #if GTP_ICS_SLOT_REPORT
         input_mt_slot(ts->input_dev, id);
         input_report_abs(ts->input_dev, ABS_MT_TRACKING_ID, id);
@@ -374,8 +377,8 @@ static void gtp_touch_down(struct goodix
         input_report_abs(ts->input_dev, ABS_MT_TOUCH_MAJOR, w);
         input_report_abs(ts->input_dev, ABS_MT_WIDTH_MAJOR, w);
 #else
-        input_report_abs(ts->input_dev, ABS_MT_POSITION_X, x);
-        input_report_abs(ts->input_dev, ABS_MT_POSITION_Y, y);
+        input_report_abs(ts->input_dev, ABS_MT_POSITION_X, y);
+        input_report_abs(ts->input_dev, ABS_MT_POSITION_Y, x);
         input_report_abs(ts->input_dev, ABS_MT_TOUCH_MAJOR, w);
         input_report_abs(ts->input_dev, ABS_MT_WIDTH_MAJOR, w);
         input_report_abs(ts->input_dev, ABS_MT_TRACKING_ID, id);
@@ -1225,8 +1228,8 @@ static struct i2c_driver goodix_ts_drive
         .probe          = goodix_ts_probe,
         .remove         = goodix_ts_remove,
 #ifndef CONFIG_HAS_EARLYSUSPEND
-//        .suspend        = goodix_ts_early_suspend,
-//        .resume         = goodix_ts_late_resume,
+        .suspend        = goodix_ts_early_suspend,
+        .resume         = goodix_ts_late_resume,
 #endif
         .id_table       = goodix_ts_id,
         .driver = {
